cpu/atmega_common/periph/rtt.c

```diff
@@ -0,0 +1,270 @@
+/*
+ * Copyright (C) 2016 Robert Hartung <hartung@ibr.cs.tu-bs.de>
+ *               2018 Matthew Blue <matthew.blue.neuro@gmail.com>
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/*
+ * @ingroup cpu_atmega_common
+ * @ingroup drivers_periph_rtt
+ * @{
+ *
+ * @file
+ * @brief       Low-level ATmega RTT driver implementation
+ *
+ * @author      Robert Hartung <hartung@ibr.cs.tu-bs.de>
+ * @author      Matthew Blue <matthew.blue.neuro@gmail.com>
+ *
+ * @}
+ */
+
+#include <avr/interrupt.h>
+
+#include "cpu.h"
+#include "irq.h"
+#include "periph/rtt.h"
+#include "periph_conf.h"
+#include "thread.h"
+
+#ifdef MODULE_XTIMER
+#include "xtimer.h"
+#endif
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+/* guard file in case no RTT device is defined */
+#if RTT_NUMOF
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#if MODULE_PERIPH_RTC
+extern void atmega_rtc_incr(void);
+#endif
+
+static inline void __asynch_wait(uint8_t num_cycles);
+
+typedef struct {
+    uint16_t ext_cnt;        /* Counter to make 8-bit timer 24-bit */
+    uint16_t ext_comp;       /* Extend compare to 24-bits */
+    rtt_cb_t alarm_cb;       /* callback called from RTT alarm */
+    void *alarm_arg;         /* argument passed to the callback */
+    rtt_cb_t overflow_cb;    /* callback called when RTT overflows */
+    void *overflow_arg;      /* argument passed to the callback */
+} rtt_state_t;
+
+static volatile rtt_state_t rtt_state;
+
+void rtt_init(void) {
+    DEBUG("Initializing RTT\n");
+
+    rtt_poweron();
+
+    /*
+     * From the datasheet section "Asynchronous Operation of Timer/Counter2"
+     * p148 for ATmega1284P.
+     * 1. Disable the Timer/Counter2 interrupts by clearing OCIE2x and TOIE2.
+     * 2. Select clock source by setting AS2 as appropriate.
+     * 3. Write new values to TCNT2, OCR2x, and TCCR2x.
+     * 4. To switch to asynchronous: Wait for TCN2UB, OCR2xUB, TCR2xUB.
+     * 5. Clear the Timer/Counter2 Interrupt Flags.
+     * 6. Enable interrupts, if needed
+    */
+
+    /* Disable all timer 2 interrupts */
+    TIMSK2 = 0;
+
+    /* Select asynchronous clock source */
+    ASSR = (1 << AS2);
+
+    /* Set counter to 0 */
+    TCNT2 = 0;
+
+    /* Reset compare values */
+    OCR2A = 0;
+    OCR2B = 0;
+
+    /* Reset timer control */
+    TCCR2A = 0;
+
+    /* 32768Hz / 1024 -> 32 ticks per second! */
+    TCCR2B = (1 << CS22) | (1 << CS21) | (1 <<  CS20);
+
+    /* Wait until not busy anymore */
+    DEBUG("RTT waits until ASSR not busy\n");
+    __asynch_wait(2);
+
+    /* Clear interrupt flags */
+    TIFR2 &= ~( (1 << OCF2B)  | (1 << OCF2A) | (1 << TOV2) );
+
+    /* Initialize callbacks */
+    rtt_state.alarm_cb = NULL;
+    rtt_state.overflow_cb = NULL;
+
+    /* Enable 8-bit overflow interrupt */
+    TIMSK2 |= (1 << TOIE2);
+
+    irq_enable();
```

Comment:
this will force enable IRQs even if they were disabled before. This is a safer alternative: 
```c
unsigned mask = irq_disable();
...
irq_restore(mask);
```

Reply:
Removed.

While the RTT requires IRQs to be enabled, it never needs to disable them globally.